### PR TITLE
[mtl] diversity and inclusion of stencils, better binding code

### LINKED ISSUE #2119: Metal scissor rectangles
Metal doesn't like the scissor rect to be outside of the render area. Thus, we need to clamp the values all the time (when starting a render pass, then within a pass upon changing the scissor).
(marked as blocker for Dota2)

### PR DESCRIPTION
**Pull request by kvark**
Fixes #2119
PR checklist:
- [x] `make` succeeds (on *nix)
- [x] `make reftests` succeeds
- [ ] tested examples with the following backends:

### CODE REVIEW

File: src/backend/metal/src/device.rs [hunk 1]
```diff
@@ -1910,7 +1915,7 @@ impl hal::Device<Backend> for Device {
 
         if swizzle != format::Swizzle::NO {
             error!("swizzling not supported");
-            return Err(image::ViewError::Unsupported);
+            //return Err(image::ViewError::Unsupported);
```
Comment (grovesNL) [inline src/backend/metal/src/device.rs]: Do we need to leave this out?

File: src/backend/metal/src/device.rs [hunk 2]
```diff
@@ -971,11 +970,17 @@ impl hal::Device<Backend> for Device {
                 .expect("no associated vertex buffer found");
             // handle wrapping offsets
             let elem_size = element.format.surface_desc().bits as pso::ElemOffset / 8;
-            let (cut_offset, base_offset) = if element.offset + elem_size <= original.stride {
+            let stride = if original.stride != 0 {
+                original.stride
+            } else {
+                warn!("Zero sized vertex stride found!");
```
Comment (grovesNL) [inline src/backend/metal/src/device.rs]: Is there a Vulkan reference for what's supposed to happen in this case? I don't see it explicitly disallowed

File: src/backend/metal/src/command.rs [hunk 3]
```diff
@@ -2160,117 +2172,50 @@ impl com::RawCommandBuffer<Backend> for CommandBuffer {
...
-                                            index: start + i,
-                                            texture: texture.as_ref().map(|&(ref texture, _)| texture.clone()),
+                                        None => (None, 0),
+                                    };
+                                    for &mut (ref loc, ref mut resources) in &mut bind_stages {
+                                        let start = layout.res_overrides[loc].buffer_id as usize;
+                                        if let Some(buffer) = buffer {
+                                            resources.add_buffer(start + i, buffer, offset as _);
                                         }
-                                    }));
```
Comment (grovesNL) [inline src/backend/metal/src/command.rs]: What happened to all the previous `commands.extend`/`push` here? Are they not actually required?

Comment (kvark) [reply src/backend/metal/src/command.rs]: The old code was doing exactly the same thing for vertex and fragment stages. The problem came from a case where a dynamic buffer is used in both, so the code would try to advance the iterator twice for the same buffer...

The new code just builds a `bind_stages` vector (of size 2 at max) beforehand, and iterates it for each descriptor set type, thus bringing the amount of code duplication to a reasonable minimum as well as resolving the dynamic buffer issue.

File: src/backend/metal/src/device.rs (hunk 2 above)
Comment (kvark) [reply src/backend/metal/src/device.rs]: Molten just considers the stride of `float4` in this case, so I thought it doesn't matter that much.
But thinking about this more (as well as checking the documentation), the 0 stride has a very clear use case: repeat the attribute. What we should be doing in Metal then is switching it to be per-instance and setting the instance rate to the max.
Perhaps, in a follow-up?

File: src/backend/metal/src/device.rs [hunk 4]
```diff
@@ -1131,8 +1139,8 @@ impl hal::Device<Backend> for Device {
                     .set_texture(Some(&at.raw));
             }
             if aspects.contains(format::Aspects::STENCIL) {
-                if let Some(mtl_format) = inner.depth_stencil {
-                    assert_eq!(mtl_format, at.mtl_format);
+                if let Some(old_format) = inner.depth_stencil {
```
[the remark this answers is not in the document]
Comment (kvark) [reply src/backend/metal/src/device.rs]: Good catch!

File: src/backend/metal/src/device.rs (hunk 1 above)
Comment (kvark) [reply src/backend/metal/src/device.rs]: I don't think swizzling should be the end of the world. We've yet to figure out how to properly limit this with VkPI group, so just commenting the return out seems fine for now.

**Comment by kvark** [comment]
Great review, thank you!
bors r=grovesNL

**Pull request merged**